### PR TITLE
Specify the utf-8 encoding when opening a text file

### PR DESCRIPTION
**Pull request by bjones1**
This is a fix for #3605, and generally for a number of places in the code where the encoding wasn't specified when opening a file. Since getting this working required more complexity than I expected, I'd appreciate @htgoebel's feedback before merging.

### CODE REVIEW

File: PyInstaller/utils/hooks/gi.py [hunk 1]
```diff
@@ -140,9 +141,10 @@ def gir_library_path_fix(path):
                          'package.', gir_file)
             return None
 
-        with open(gir_file, 'r') as f:
+        with open_file(gir_file, text_read_mode, encoding='utf-8') as f:
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/gi.py]: How can we be sure this is utf-8 encoded? If you have some source for this assumption, please state this in a comment.

File: PyInstaller/utils/hooks/gi.py [hunk 2]
```diff
@@ -140,9 +141,10 @@ def gir_library_path_fix(path):
...
             lines = f.readlines()
-        with open(os.path.join(CONF['workpath'], gir_name), 'w') as f:
+        with open_file(os.path.join(CONF['workpath'], gir_name), 'w',
+                  encoding='utf-8') as f:
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/gi.py]: How can we be sure utf-8 is expected or accepted? If you have some source for this assumption, please state this in a comment.

File: news/3605.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Specify an encoding of UTF-8 when opening text files.
```
Comment (htgoebel) [inline news/3605.bugfix.rst]: Even when rewording after the changes I'm requesting, this should state that we basically open *all* text files using  utf-8 *explicitly* and do so for both reading and writing

File: tests/functional/test_updating_manifest.py [hunk 4]
```diff
@@ -50,6 +49,10 @@ def test_reading_manifest(tmpdir):
     XML manifest files are written with UTF-8 even localized Windows.
     We check here not to use local encoding against UTF-8 manifest file.
     """
+    # This import only works on Windows. Place it here, protected by the
+    # ``@skipif_notwin`` decorator.
+    from PyInstaller.utils.win32 import winmanifest
+
```
Comment (htgoebel) [inline tests/functional/test_updating_manifest.py]: Unrelated change slipped in. Looks like you need to rebase :-)

File: PyInstaller/building/build_main.py [hunk 5]
```diff
@@ -219,8 +220,8 @@ def __init__(self, scripts, pathex=None, binaries=None, datas=None,
             # Create a Python module which contains the decryption key which will
             # be used at runtime by pyi_crypto.PyiBlockCipher.
             pyi_crypto_key_path = os.path.join(CONF['workpath'], 'pyimod00_crypto_key.py')
-            with open(pyi_crypto_key_path, 'w') as f:
-                f.write('key = %r\n' % cipher.key)
+            with open_file(pyi_crypto_key_path, 'w', encoding='utf-8') as f:
```
Comment (htgoebel) [inline PyInstaller/building/build_main.py]: Okay, but not necessary since the key will be `repr`-ed and thus be ascii-only.
Also when using an encoding, the first line in the file must be
```
# -*- coding: utf-8 -*-
```

File: PyInstaller/building/build_main.py [hunk 6]
```diff
@@ -29,7 +29,8 @@
 from .. import compat
 from .. import log as logging
 from ..utils.misc import absnormpath, compile_py_files
-from ..compat import is_py2, is_win, PYDYLIB_NAMES, VALID_MODULE_TYPES
+from ..compat import is_py2, is_win, PYDYLIB_NAMES, VALID_MODULE_TYPES, \
+    open_file, text_read_mode, text_type, unicode_writer
```
Comment (htgoebel) [inline PyInstaller/building/build_main.py]: Just an idea: If we `import open_file as open`, we will save quite some not-meaningful changes when dropping py2-compatilitiy. What do you think?

File: PyInstaller/building/build_main.py [hunk 7]
```diff
@@ -219,8 +220,8 @@ def __init__(self, scripts, pathex=None, binaries=None, datas=None,
             # Create a Python module which contains the decryption key which will
             # be used at runtime by pyi_crypto.PyiBlockCipher.
             pyi_crypto_key_path = os.path.join(CONF['workpath'], 'pyimod00_crypto_key.py')
-            with open(pyi_crypto_key_path, 'w') as f:
-                f.write('key = %r\n' % cipher.key)
+            with open_file(pyi_crypto_key_path, 'w', encoding='utf-8') as f:
+                f.write(text_type('key = %r\n' % cipher.key))
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/build_main.py]: Because, in Python 2, `type('key = %r\n' % cipher.key)` is `<type 'str'>`, not `<type 'unicode'>`. That is, in Python 2, the *default* type is `binary_type`, not `text_type`. **Aaargh, Python 2 is so painful for Unicode!**

We could use `u'key = %r\n' % cipher.key`, but this makes removing this later (when we drop Python 2 support) harder.

File: PyInstaller/building/build_main.py [hunk 8]
```diff
@@ -618,28 +619,29 @@ def dependency_description(name, depInfo):
...
-            self.graph.create_xref(fh)
+        with open_file(CONF['xref-file'], 'w', encoding='utf-8') as fh:
+            self.graph.create_xref(unicode_writer(fh))
             logger.info("Graph cross-reference written to %s", CONF['xref-file'])
         if logger.getEffectiveLevel() > logging.DEBUG:
             return
-        with open(CONF['dot-file'], 'w') as fh:
-            self.graph.graphreport(fh)
+        with open_file(CONF['dot-file'], 'w', encoding='utf-8') as fh:
+            self.graph.graphreport(unicode_writer(fh))
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/build_main.py]: I don't see a template, but simply a bunch of text inside [itergraphreport](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/lib/modulegraph/modulegraph.py#L3224). Were you thinking of adding an extra line there to output the charset?

File: PyInstaller/building/build_main.py [hunk 9]
```diff
@@ -618,28 +619,29 @@ def dependency_description(name, depInfo):
...
 
     def _write_graph_debug(self):
         """Write a xref (in html) and with `--log-level DEBUG` a dot-drawing
         of the graph.
         """
         from ..config import CONF
-        with open(CONF['xref-file'], 'w') as fh:
-            self.graph.create_xref(fh)
+        with open_file(CONF['xref-file'], 'w', encoding='utf-8') as fh:
+            self.graph.create_xref(unicode_writer(fh))
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/build_main.py]: I agree. I'll work on that.

File: PyInstaller/building/build_main.py [hunk 10]
```diff
@@ -778,10 +780,16 @@ def TkTree(*args, **kwargs):
...
+    # Execute the specfile.
+    if is_py2:
+        # Deal with encoding annoyances in Python 2.
+        execfile(spec, spec_namespace)
+    else:
+        with open_file(spec, text_read_mode, encoding='utf-8') as f:
+            # Associate file names with the code object, to make tracebacks
+            # more informative.
+            code = compile(f.read(), spec, 'exec')
+            exec(code, spec_namespace)
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/build_main.py]: Thanks for looking. Yes, you're correct -- we need to let Python figure out its own source file encoding, and follow that the standard library does. I'd suggest not adding `source_read-mode` if this is the only use of it, though.

Python 3's `exec` function only operates on strings, not bytes. However, `compile` accepts both, hence its use here.

How about:

``` Python
    # Executing the specfile.
    with open(spec, 'rb' if is_py3 else 'rU') as f:
        code = compile(f.read(), spec, 'exec')
    exec(code, spec_namespace)
```

File: PyInstaller/building/makespec.py [hunk 11]
```diff
@@ -434,16 +434,16 @@ def main(scripts, name=None, onefile=None,
 
     # Write down .spec file to filesystem.
     specfnm = os.path.join(specpath, name + '.spec')
-    with open(specfnm, 'w') as specfile:
+    with open_file(specfnm, 'w', encoding='utf-8') as specfile:
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/makespec.py]: Good point. Done.

File: PyInstaller/building/makespec.py [hunk 12]
```diff
@@ -434,16 +434,16 @@ def main(scripts, name=None, onefile=None,
...
         if onefile:
-            specfile.write(onefiletmplt % d)
+            specfile.write(text_type(onefiletmplt % d))
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/makespec.py]: As before, Python 2's default type is `str`, which is `binary_type`, not `text_type`. So, this is actually necessary.

File: PyInstaller/compat.py [hunk 13]
```diff
@@ -99,6 +99,37 @@
...
+
+# These are copied from ``six``.
+#
+# Type for representing (Unicode) textual data.
+text_type = unicode if is_py2 else str
+# Type for representing binary data.
+binary_type = str if is_py2 else bytes
+
+
+# Create a class which converts all writes to unicode first. For use with
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/compat.py]: Done.

File: PyInstaller/hooks/hook-_tkinter.py [hunk 14]
```diff
@@ -77,7 +78,7 @@ def _warn_if_activetcl_or_teapot_installed(tcl_root, tcltree):
 
     mentions_activetcl = False
     mentions_teapot = False
-    with open(init_resource, 'r') as init_file:
+    with open_file(init_resource, text_read_mode, encoding='utf-8') as init_file:
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/hooks/hook-_tkinter.py]: Good point. I don't know the encoding. The existing code assumes an encoding: ASCII for Python 2, and `locale.getpreferredencoding()` for Python 3. While this has (hopefully) worked in the past, it seems poor practice for the future. Do you have any thoughts? I think we should explicitly specify an encoding, rather than relying on this "default" which is Python- and platform-dependent.

File: PyInstaller/hooks/hook-sqlalchemy.py [hunk 15]
```diff
@@ -59,7 +60,7 @@ def hook(hook_api):
         if isinstance(node, SourceModule) and \
                 node.identifier.startswith('sqlalchemy.'):
             known_imports.add(node.identifier)
-            with open(node.filename) as f:
+            with open_file(node.filename, text_read_mode, encoding='utf-8') as f:
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/hooks/hook-sqlalchemy.py]: Aargh. I want to avoid guessing encodings, unless there's a stdlib function to do this. I have no idea what to do here.

File: PyInstaller/utils/hooks/gi.py (hunk 1 above)
Comment (bjones1) [reply PyInstaller/utils/hooks/gi.py]: As before, let's specify and encoding, instead of relying on a poorly-documented "default".

File: PyInstaller/utils/hooks/gi.py (hunk 2 above)
Comment (bjones1) [reply PyInstaller/utils/hooks/gi.py]: As before, let's specify and encoding, instead of relying on a poorly-documented "default".

File: tests/functional/test_updating_manifest.py (hunk 4 above)
Comment (bjones1) [reply tests/functional/test_updating_manifest.py]: Should be fixed already -- perhaps this is github's confusion.

File: news/3605.bugfix.rst (hunk 3 above)
Comment (bjones1) [reply news/3605.bugfix.rst]: I agree, but that also depends on how we'll handle the opening of files in the hooks. I'll update the comment for now.

File: PyInstaller/building/build_main.py (hunk 5 above)
Comment (bjones1) [reply PyInstaller/building/build_main.py]: For consistence, I'd suggest sticking with UTF-8. I've added the comment.

File: PyInstaller/building/build_main.py (hunk 6 above)
Comment (bjones1) [reply PyInstaller/building/build_main.py]: I'm ambivalent. As you say, this makes moving to Python 3 only easier. But, I don't like the idea of re-defining a built-in function.

**Comment by bjones1** [comment]
Thanks for the thorough review. Getting Unicode right is very, very hard. Especially when trying to make it work with Python 2 and Python 3.

File: PyInstaller/hooks/hook-_tkinter.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (htgoebel) [reply PyInstaller/hooks/hook-_tkinter.py]: Well, explicit is better than implicit :-) . But when reading https://www.tcl.tk/doc/howto/i18n.html#system_encoding, I assume `locale.getpreferredencoding()` is the right thing to do. I suggest using this explicitly and refering to the mentioned url.

File: PyInstaller/hooks/hook-sqlalchemy.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (htgoebel) [reply PyInstaller/hooks/hook-sqlalchemy.py]: There is a stdlib function, but not in Python2. This is hwy it is copied to `modulegraph.util`. This function does not exactly "guess", but looks whether there is a `coding =` header.

File: PyInstaller/utils/hooks/gi.py (hunk 1 above)
Comment (htgoebel) [reply PyInstaller/utils/hooks/gi.py]: Looks like git files are XML https://developer.gnome.org/gi/stable/gi-gir-reference.html. Thus strictly speaking we should use an xml parser which cares about the correct encoding. But since utf-8 is the XML default encoding, we can life with this. Please add a respective commend.

File: PyInstaller/building/build_main.py [hunk 16]
```diff
@@ -219,8 +220,9 @@ def __init__(self, scripts, pathex=None, binaries=None, datas=None,
             # Create a Python module which contains the decryption key which will
             # be used at runtime by pyi_crypto.PyiBlockCipher.
             pyi_crypto_key_path = os.path.join(CONF['workpath'], 'pyimod00_crypto_key.py')
-            with open(pyi_crypto_key_path, 'w') as f:
-                f.write('key = %r\n' % cipher.key)
+            with open_file(pyi_crypto_key_path, 'w', encoding='utf-8') as f:
+                f.write(text_type('# -*- coding: utf-8 -*-\nkey = %r\n' %
```
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/build_main.py]: Done.

File: PyInstaller/hooks/hook-_tkinter.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/hooks/hook-_tkinter.py]: Great! Done.

File: PyInstaller/hooks/hook-sqlalchemy.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/hooks/hook-sqlalchemy.py]: I prefer to let Python determine source file encoding (in `build_main`) where possible, so I'll simply import and use the ModuleGraph `guess_encoding` function.

File: PyInstaller/utils/hooks/gi.py (hunk 1 above)
Comment (bjones1) [reply PyInstaller/utils/hooks/gi.py]: Done, thanks.

File: PyInstaller/building/build_main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bjones1) [reply PyInstaller/building/build_main.py]: This is done.

**Comment by tangb** [comment]
Hello

Has fix for that issue been released ?
I'm encountering the same error using latest version 3.5 under win10 and I don't find any infos about it in officiel changelog https://pyinstaller.readthedocs.io/en/stable/CHANGES.html.

Thank you for your answer and for this amazing project :relaxed:

**Comment by bjones1** [comment]
Yes. It's released and in the changlog (search for 3605).

**Comment by tangb** [comment]
Ok.

The issue I'm encountering is located here: https://github.com/pyinstaller/pyinstaller/blob/master/PyInstaller/compat.py#L427 

There is a discussion on stackoverflow https://stackoverflow.com/questions/47692960/error-when-using-pyinstaller-unicodedecodeerror-utf-8-codec-cant-decode-byt and they were waiting for a fix related to this issue (and 3605)

Do I create a new issue about it ?

**Comment by bjones1** [comment]
Yes.

**Comment by santoslopes** [comment]
Not solved

**Comment by NotVeryGoodAtCodingButOk** [comment]
True, this video is still failing. Please pay attention to this problem!

**Comment by stefaneidelloth** [comment]
Related SO-Question pointing to this issue ticket:

https://stackoverflow.com/questions/47692960/error-when-using-pyinstaller-unicodedecodeerror-utf-8-codec-cant-decode-byt

If there is a newer ticket you might want to link it.

**Comment by AndreiGurko** [comment]
In PyInstaller version == 4.7 the same problem raised again

**Comment by bwoodsend** [comment]
Create a new issue. There's no point posting here.
